localnodedns/controllers/localnodedns_controller.go

```diff
@@ -0,0 +1,117 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/addon"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/addon/pkg/status"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	api "sigs.k8s.io/cluster-addons/localnodedns/api/v1alpha1"
+)
+
+var _ reconcile.Reconciler = &LocalNodeDNSReconciler{}
+
+// LocalNodeDNSReconciler reconciles a LocalNodeDNS object
+type LocalNodeDNSReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	declarative.Reconciler
+	watchLabels declarative.LabelMaker
+}
+
+func (r *LocalNodeDNSReconciler) setupReconciler(mgr ctrl.Manager) error {
+	addon.Init()
+
+	labels := map[string]string{
+		"k8s-app": "localnodedns",
+	}
+
+	r.watchLabels = declarative.SourceLabel(mgr.GetScheme())
+
+	return r.Reconciler.Init(mgr, &api.LocalNodeDNS{},
+		declarative.WithRawManifestOperation(replaceVariables(mgr)),
+		declarative.WithObjectTransform(declarative.AddLabels(labels)),
+		declarative.WithOwner(declarative.SourceAsOwner),
+		declarative.WithLabels(r.watchLabels),
+		declarative.WithStatus(status.NewBasic(mgr.GetClient())),
+		declarative.WithObjectTransform(addon.TransformApplicationFromStatus),
+		declarative.WithManagedApplication(r.watchLabels),
+		declarative.WithObjectTransform(addon.ApplyPatches),
+	)
+}
+
+// +kubebuilder:rbac:groups=addons.x-k8s.io,resources=localnodedns,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=addons.x-k8s.io,resources=localnodedns/status,verbs=get;update;patch
+
+func (r *LocalNodeDNSReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := r.setupReconciler(mgr); err != nil {
+		return err
+	}
+
+	c, err := controller.New("localnodedns-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to LocalNodeDNS
+	err = c.Watch(&source.Kind{Type: &api.LocalNodeDNS{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to deployed objects
+	_, err = declarative.WatchAll(mgr.GetConfig(), c, r, r.watchLabels)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func replaceVariables(mgr ctrl.Manager) declarative.ManifestOperation {
+	return func(ctx context.Context, object declarative.DeclarativeObject, s string) (string, error) {
+		o := object.(*api.LocalNodeDNS)
+		kubeProxyMode, err := findKubeProxyMode(ctx, mgr.GetClient())
+		if err != nil {
+			fmt.Println("Error determining kube-proxy mode: Defaulting to iptables")
+		}
+
+		if o.Spec.DNSDomain == "" {
+			o.Spec.DNSDomain = "cluster.local"
+		}
+
+		if o.Spec.DNSIP == "" {
+			o.Spec.DNSIP = "169.254.20.10"
+		}
+
+		if o.Spec.ClusterIP == "" {
```

Comment:
There are similar functions in the coredns-operators for finding the ClusterIP and DNS Domain for a cluster. Maybe we could reuse it? Wdyt?

Reply:
Yes, we should start to create a shared library of these functions.  I think probably in kubebuilder-declarative-pattern, because if e.g. the DNS team takes on ownership of this operator this repo may eventually end up pretty empty :-)

I think we can do this in a follow on PR, if you'd like.  (You can add a TODO here so we don't forget)